Bug 1417659: L10N screenshot for tracking protection long-press-reload

garvankeeley opened the pull request:
https://bugzilla.mozilla.org/show_bug.cgi?id=1417659

ContentBlockerHelper.swift and ContentBlockerSettingsViewController.swift needed to be updated for screengraph usage.

garvankeeley commented on Client/Frontend/ContentBlocker/ContentBlockerHelper.swift:
```diff
@@ -55,6 +55,17 @@ class ContentBlockerHelper {
+        static func accessibilityId(for state: EnabledState) -> String {
+            switch state {
+            case .on:
+                return "Settings.TrackingProtectionOption.OnLabel"
```
IMO this should be `Strings.trackingProtectionAccessibilityId.onLabel`, otherwise this string gets repeated in 3 places, 1) Strings.swift, 2) here, 3) when used in tests.
But there is no precedent in the code for this method.

Then tests would be able to include Strings.swift, and use constants for accessibility identifiers.

jhugman replied:
I'd like to come up with some standard for `accessibilityId`.

garvankeeley commented on XCUITests/FxScreenGraph.swift:
```diff
@@ -540,6 +561,12 @@ func createScreenGraph(for test: XCTestCase, with app: XCUIApplication) -> Scree
         }
+
+        let trackingProtectionButton = app.sheets.element(boundBy: 0).buttons.element(boundBy: 1)
+        let tpIsOn = "trackingProtectionSetting == \(FxUserState.TrackingProtectionSetting.alwaysOn.rawValue) || (trackingProtectionSetting == \(FxUserState.TrackingProtectionSetting.privateOnly.rawValue) && isPrivate == YES)"
```
Maybe a better way? NSPredicates trip me up (in this case due to the use of Enum) and this is what I end up with.

garvankeeley replied:
Do you mean a) FxUserState gets a member `let trackingProtectionOnCondition = "the predicate string"`? Makes sense to encapsulate it there and make the enum private yeah.

And for b) I don't know where else it would go (only userstate seems to be the place to put it).

jhugman replied:
a. Hadn't thought of putting the predicate into the enum value itself. Not sure if I feel dirty doing that :O) I was only really talking about making the enum fileprivate.
b. If it was private, it could go right next to the `FxUserState`. It wouldn't be exposed to the tests, (or anything outside of `FxScreenGraph`). Anything to make it less verbose.

garvankeeley replied:
Much nicer making the enum fileprivate at top-level, updating and then landing, thanks!

garvankeeley commented on Client/Frontend/ContentBlocker/ContentBlockerSettingViewController.swift:
```diff
@@ -61,7 +61,8 @@ class ContentBlockerSettingViewController: ContentBlockerSettingsTableView {
         let enabledSetting: [CheckmarkSetting] = EnabledStates.map { option in
-            return CheckmarkSetting(title: NSAttributedString(string: option.settingTitle), subtitle: nil, isEnabled: {
+            let id = ContentBlockerHelper.EnabledState.accessibilityId(for: option)
+            return CheckmarkSetting(title: NSAttributedString(string: option.settingTitle), subtitle: nil, accessibilityIdentifier: id, isEnabled: {
```
In order for the screengraph to change the Settings, I needed proper  accessibilityIdentifier set on the options (I wasn't aware when I landed this a while ago of the implications of not doing that, now I know, do it the first time).

jhugman commented on L10nSnapshotTests/L10nSnapshotTests.swift:
```diff
@@ -134,14 +134,26 @@ class L10nSnapshotTests: L10nBaseSnapshotTests {
     }
 
     func test13ReloadButtonContextMenu() {
+        navigator.performAction(Action.ToggleTrackingProtectionSettingAlwaysOn)
```
Tip: use 

```swift
navigator.toggleOn(userState.trackingProtectionSetting == FxUserState.TrackingProtectionSetting.alwaysOn.rawValue, withAction: Action.ToggleTrackingProtectionSettingAlwaysOn)
```

This would make it tolerant to us changing the defaults in the future, but… considerably ugly in the present.

jhugman commented on L10nSnapshotTests/L10nSnapshotTests.swift:
```diff
@@ -134,14 +134,26 @@ class L10nSnapshotTests: L10nBaseSnapshotTests {
         snapshot("13ContextMenuReloadButton-02", waitForLoadingIndicator: false)
-        navigator.back()
+
+        navigator.performAction(Action.ToggleTrackingProtectionPerTabEnabled)
```
Use 

```swift
navigator.toggleOff(userState.trackingProtectionPerTabEnabled, withAction: Action.ToggleTrackingProtectionPerTabEnabled)
```

garvankeeley commented:
Thanks for review, hopefully I get to do more screengraph stuff to keep this knowledge fresh.
I think I addressed all comments in last push (could you take a look quickly?), I just need to clarify https://github.com/mozilla-mobile/firefox-ios/pull/3461#discussion_r151767646
